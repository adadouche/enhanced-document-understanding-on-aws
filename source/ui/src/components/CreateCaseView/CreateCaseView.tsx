/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the License). You may not use this file except in compliance      *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/
import { AppLayout, ContentLayout, SpaceBetween } from '@cloudscape-design/components';
import { API } from 'aws-amplify';
import React from 'react';
import { useNavigate } from 'react-router-dom';
import { API_NAME, MAX_CASE_NAME_LENGTH, MIN_CASE_NAME_LENGTH } from '../../utils/constants';
import { generateToken } from '../DocumentTable/DocumentTable';
import { FormContent, FormHeader } from './form-content';

type CreateCaseViewProps = {};
let token: string;

export default function CreateCaseView(props: CreateCaseViewProps) {
    const [caseNameError, setCaseNameError] = React.useState('');
    const [currentStatus, setCurrentStatus] = React.useState('');
    const navigate = useNavigate();

    const postResource = async (endpoint: string, params = {}) => {
        try {
            setCurrentStatus('loading');
            token = await generateToken();
            const response = await API.post(API_NAME, endpoint, {
                headers: {
                    Authorization: token
                },
                body: params
            });
            setCurrentStatus('success');
            setTimeout(function () {
                navigate('/');
            }, 1000);
            return response;
        } catch (err) {
            setCurrentStatus('error');
            console.error(err);
        }
    };

    const handleButtonClick = async (caseName: string) => {
        if (!caseName.match(`^[a-zA-Z0-9_ -]{${MIN_CASE_NAME_LENGTH},${MAX_CASE_NAME_LENGTH}}$`)) {
            setCaseNameError(
                'Case name can only include alphanumeric characters, -, _, and spaces and must be between ' +
                    MIN_CASE_NAME_LENGTH +
                    ' and ' +
                    MAX_CASE_NAME_LENGTH +
                    ' characters.'
            );
        } else {
            setCurrentStatus('loading');
            setCaseNameError('');
            await postResource(`case`, { caseName: caseName });
        }
    };

    return (
        <div>
            <AppLayout
                contentType="form"
                content={
                    <ContentLayout
                        header={
                            <SpaceBetween size="m">
                                <FormHeader />
                            </SpaceBetween>
                        }
                        data-testid="create-case-view-contentlayout"
                    >
                        <FormContent
                            handleButtonClick={handleButtonClick}
                            caseNameError={caseNameError}
                            currentStatus={currentStatus}
                        />
                    </ContentLayout>
                }
                headerSelector="#header"
                navigationHide
                toolsHide
                date-testid="create-case-view-applayout"
            />
        </div>
    );
}
