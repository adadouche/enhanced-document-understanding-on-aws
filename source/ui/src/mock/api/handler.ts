// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { rest } from 'msw';

export const MOCK_CONFIG = {
    ApiEndpoint: 'https://jsonplaceholder.typicode.com/',
    AWS_REGION: 'us-east-1',
    USER_POOL_ID: 'test_user_pool_id',
    USER_POOL_CLIENT_ID: 'testClientId',
    DATA_BUCKET_NAME: 'test-bucket'
};

export const getSignedUrlResponse = {
    downloadUrl: `${MOCK_CONFIG.ApiEndpoint}fake-url`
};

const getDocumentResponse = {
    key: `${MOCK_CONFIG.ApiEndpoint}/fake-key`
};

const getCasesResponse = {
    Count: 4,
    ScannedCount: 4,
    Items: [
        {
            CASE_ID: { S: 'fake-case-id-1' },
            CASE_NAME: { S: 'fake-case-name-1' },
            DOCUMENT_ID: { S: '0000' },
            STATUS: { S: 'success' },
            USER_ID: { S: 'fake-user-id' },
            DOC_COUNT: { N: '1' },
            CREATION_TIMESTAMP: { 'S': '2000-01-01T00:00:00.000Z' }
        },
        {
            CASE_ID: { S: 'fake-case-id-2' },
            CASE_NAME: { S: 'fake-case-name-2' },
            DOCUMENT_ID: { S: '0000' },
            STATUS: { S: 'success' },
            USER_ID: { S: 'fake-user-id' },
            DOC_COUNT: { N: '0' },
            CREATION_TIMESTAMP: { 'S': '2000-01-01T00:00:00.000Z' }
        },
        {
            CASE_ID: { S: 'fake-case-id-3' },
            CASE_NAME: { S: 'fake-case-name-3' },
            DOCUMENT_ID: { S: '0000' },
            STATUS: { S: 'success' },
            USER_ID: { S: 'fake-user-id' },
            DOC_COUNT: { N: '0' },
            CREATION_TIMESTAMP: { 'S': '2000-01-01T00:00:00.000Z' }
        },
        {
            CASE_ID: { S: 'fake-case-id-4' },
            CASE_NAME: { S: 'fake-case-name-4' },
            DOCUMENT_ID: { S: '0000' },
            STATUS: { S: 'success' },
            USER_ID: { S: 'fake-user-id' },
            DOC_COUNT: { N: '0' },
            CREATION_TIMESTAMP: { 'S': '2000-01-01T00:00:00.000Z' }
        }
    ]
};

const getDocumentByCaseIdResponse = {
    Count: 1,
    ScannedCount: 1,
    Items: [
        {
            BUCKET_NAME: { S: 'fake-bucket-name-1' },
            CASE_ID: { S: 'fake-case-id-1' },
            CASE_NAME: { S: 'fake-case-name-1' },
            DOCUMENT_ID: { S: 'fake-doc-id1' },
            DOCUMENT_TYPE: { S: 'passport' },
            S3_KEY: {
                S: 'fake-s3-key.jpg'
            },
            UPLOADED_FILE_EXTENSION: { S: 'jpg' },
            UPLOADED_FILE_NAME: { S: 'fake-file-name' },
            USER_ID: { S: 'fake-user-id' },
            CREATION_TIMESTAMP: { 'S': '2000-01-01T00:00:00.000Z' }
        }
    ]
};

export const handlers = [
    rest.get('/runtimeConfig.json', (_, res, ctx) => {
        return res(ctx.status(200), ctx.text(JSON.stringify(MOCK_CONFIG)));
    }),
    rest.get(`${MOCK_CONFIG.ApiEndpoint}cases`, (req, res, ctx) => {
        let response: any = getCasesResponse;
        if (req.params) {
            response = {
                ...getCasesResponse,
                LastEvaluatedKey: true
            };
        }
        return res(ctx.status(200), ctx.text(JSON.stringify(response)));
    }),
    rest.get(`${MOCK_CONFIG.ApiEndpoint}case/:caseId`, (_, res, ctx) => {
        return res(ctx.status(200), ctx.text(JSON.stringify(getDocumentByCaseIdResponse)));
    }),
    rest.get(`${MOCK_CONFIG.ApiEndpoint}document/:caseId/:documentId`, (req, res, ctx) => {
        return res(ctx.status(200), ctx.text(JSON.stringify(getDocumentResponse)));
    }),
    rest.get(`${MOCK_CONFIG.ApiEndpoint}document/download`, (_, res, ctx) => {
        return res(ctx.status(200), ctx.text(JSON.stringify(getSignedUrlResponse)));
    }),
    rest.get(`${MOCK_CONFIG.ApiEndpoint}inferences/:caseId/:documentId`, (_, res, ctx) => {
        return res(ctx.status(200), ctx.text(JSON.stringify(['textract-detectText', 'entity-medical-locations'])));
    }),
    rest.get(`${MOCK_CONFIG.ApiEndpoint}inferences/:caseId/:documentId/:inferenceType`, (req, res, ctx) => {
        return res(ctx.status(200), ctx.text(JSON.stringify({})));
    }),
    rest.post(`${MOCK_CONFIG.ApiEndpoint}document`, (_, res, ctx) => {
        return res(
            ctx.status(200),
            ctx.text(
                JSON.stringify({
                    fields: [
                        {
                            key: 'test-entry'
                        }
                    ],
                    url: 'signed-url'
                })
            )
        );
    }),
    rest.all('/signed-url', (req, res, ctx) => {
        return res(ctx.status(403));
    })
];
