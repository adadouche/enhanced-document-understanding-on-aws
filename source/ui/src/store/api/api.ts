/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the License). You may not use this file except in compliance      *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { BaseQueryFn, createApi, FetchArgs, fetchBaseQuery, FetchBaseQueryError } from '@reduxjs/toolkit/query/react';
import { generateToken, getRuntimeConfig } from '../../utils/apiHelper';

export const dynamicBaseQuery: BaseQueryFn<string | FetchArgs, unknown, FetchBaseQueryError> = async (
    args: any,
    api: any,
    extraOptions: any
) => {
    const runtimeConfig = await getRuntimeConfig();
    const rawBaseQuery = fetchBaseQuery({
        baseUrl: runtimeConfig.ApiEndpoint,
        prepareHeaders: async (headers: any) => {
            if (process.env.NODE_ENV !== 'test') {
                const token = await generateToken();
                headers.set('Authorization', `${token}`);
            }
            return headers;
        }
    });
    return rawBaseQuery(args, api, extraOptions);
};
export const api = createApi({
    reducerPath: 'api',
    baseQuery: dynamicBaseQuery,
    tagTypes: ['Cases'],
    endpoints: () => ({
       
    }),
    refetchOnMountOrArgChange: true
});

