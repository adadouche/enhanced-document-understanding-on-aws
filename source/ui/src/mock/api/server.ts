// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { handlers } from "./handler";
import { setupServer } from "msw/node";
export const server = setupServer(...handlers);