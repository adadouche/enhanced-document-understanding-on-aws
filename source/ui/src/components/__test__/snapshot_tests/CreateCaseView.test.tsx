/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import '@testing-library/jest-dom';
import { API } from '@aws-amplify/api';
import CreateCaseView from '../../CreateCaseView/CreateCaseView';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import renderer from 'react-test-renderer';

const mockAPI = {
    get: jest.fn(),
    post: jest.fn()
};
jest.mock('@aws-amplify/api');
API.get = mockAPI.get;
API.post = mockAPI.post;

beforeEach(() => {
    mockAPI.post.mockReset();
});

jest.mock('@cloudscape-design/components', () => {
    const Components = jest.genMockFromModule('@cloudscape-design/components') as any;
    for (const componentName of Object.keys(Components)) {
        Components[componentName] = componentName;
    }
    return Components;
});

test('Snapshot test', async () => {
    const tree = renderer
        .create(
            <MemoryRouter initialEntries={['/createCase']}>
                <Routes>
                    <Route path="/createCase" element={<CreateCaseView />} />
                </Routes>
            </MemoryRouter>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
