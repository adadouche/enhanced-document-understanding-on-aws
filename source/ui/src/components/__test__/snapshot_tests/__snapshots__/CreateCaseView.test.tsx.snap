// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
<div>
  <AppLayout
    content={
      <ContentLayout
        data-testid="create-case-view-contentlayout"
        header={
          <SpaceBetween
            size="m"
          >
            <FormHeader />
          </SpaceBetween>
        }
      >
        <FormContent
          caseNameError=""
          currentStatus=""
          handleButtonClick={[Function]}
        />
      </ContentLayout>
    }
    contentType="form"
    date-testid="create-case-view-applayout"
    headerSelector="#header"
    navigationHide={true}
    toolsHide={true}
  />
</div>
`;
